File: src/adb.ts
import { access, mkdir, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { findExecutable, runProcess } from "./process.js";

export class AdbError extends Error {}

const packagePattern = /^[A-Za-z][A-Za-z0-9_]*(?:\.[A-Za-z0-9_]+)+$/;
const pngSignature = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);

export function extractPng(output: Uint8Array): Buffer {
  const buffer = Buffer.from(output);
  const offset = buffer.indexOf(pngSignature);
  if (offset < 0) throw new AdbError("screencap did not return a PNG image");
  return buffer.subarray(offset);
}

function assertPackageName(packageName: string): void {
  if (!packagePattern.test(packageName)) throw new AdbError(`invalid package: ${packageName}`);
}

export class Adb {
  readonly serial: string;

  private constructor(
    readonly host: string,
    readonly executable: string,
    readonly port = 5555,
  ) {
    this.serial = `${host}:${port}`;
  }

  static async create(host: string, port = 5555): Promise<Adb> {
    const executable = await findExecutable("adb");
    if (!executable) {
      throw new AdbError(
        "adb was not found; install Android Platform Tools " +
          "(macOS: brew install --cask android-platform-tools)",
      );
    }
    return new Adb(host, executable, port);
  }

  private async run(args: string[]): Promise<string> {
    const result = await runProcess(this.executable, ["-s", this.serial, ...args]);
    return result.stdout.toString().trim();
  }

  async connect(): Promise<void> {
    const result = await runProcess(this.executable, ["connect", this.serial], {
      allowFailure: true,
    });
    const output = `${result.stdout}\n${result.stderr}`.toLowerCase();
    if (result.code !== 0 || output.includes("unable") || output.includes("failed")) {
      throw new AdbError(output.trim());
    }
  }

  async shell(command: string): Promise<string> {
    await this.connect();
    return this.run(["shell", command]);
  }

  // Send a keyboard, mouse, or touch event through the on-device "input" tool.
  // Args follow "input <source> <command> <arg>...", e.g. ["keyevent", "KEYCODE_DPAD_OK"],
  // ["mouse", "tap", "500", "500"], ["keyboard", "text", "hello"].
  async input(args: readonly string[]): Promise<string> {
    if (args.length === 0) throw new AdbError("input requires an event (e.g. keyevent KEYCODE_DPAD_OK)");
    for (const arg of args) {
      // Reject shell metacharacters: the event string is interpolated into an
      // `adb shell` command, so a space, quote, or $ would break out of it.
      if (!/^[A-Za-z0-9_,.+%@:=|-]+$/.test(arg)) {
        throw new AdbError(`unsafe input argument: ${arg}`);
      }
    }
    await this.connect();
    return this.run(["shell", `input ${args.join(" ")}`]);
  }

  async info(): Promise<Record<string, string>> {
    const properties = [
      ["model", "ro.product.model"],
      ["android", "ro.build.version.release"],
      ["sdk", "ro.build.version.sdk"],
      ["abis", "ro.product.cpu.abilist"],
    ] as const;
    return Object.fromEntries(
      await Promise.all(
        properties.map(async ([name, property]) => [name, await this.shell(`getprop ${property}`)]),
      ),
    );
  }

  currentApp(): Promise<string> {
    return this.shell("dumpsys activity activities | grep -m 1 mResumedActivity");
  }

  audio(): Promise<string> {
    return this.shell("dumpsys audio | grep -E -m 8 'STREAM_MUSIC|Current:|Devices:'");
  }

  async packages(query?: string): Promise<string[]> {
    const packages = (await this.shell("pm list packages"))
      .split("\n")
      .map((line) => line.replace(/^package:/, ""))
      .filter(Boolean);
    const filtered = query
      ? packages.filter((name) => name.toLowerCase().includes(query.toLowerCase()))
      : packages;
    return filtered.sort();
  }

  async install(files: string[], replaceExisting = true): Promise<string> {
    if (files.length === 0) throw new AdbError("at least one APK is required");
    for (const file of files) await access(file);
    await this.connect();
    const command = files.length > 1 ? "install-multiple" : "install";
    return this.run([command, ...(replaceExisting ? ["-r"] : []), ...files.map((file) => resolve(file))]);
  }

  async uninstall(packageName: string, keepData = false): Promise<string> {
    assertPackageName(packageName);
    await this.connect();
    return this.run(["uninstall", ...(keepData ? ["-k"] : []), packageName]);
  }

  async isPackageInstalled(packageName: string): Promise<boolean> {
    assertPackageName(packageName);
    return (await this.shell(`pm path ${packageName}`)).split("\n").some((line) =>
      line.startsWith("package:"),
    );
  }

  forceStop(packageName: string): Promise<string> {
    assertPackageName(packageName);
    return this.shell(`am force-stop ${packageName}`);
  }

  launch(packageName: string): Promise<string> {
    assertPackageName(packageName);
    return this.shell(`monkey -p ${packageName} -c android.intent.category.LAUNCHER 1`);
  }

  async screenshot(destination: string): Promise<string> {
    await this.connect();
    const output = resolve(destination);
    await mkdir(resolve(output, ".."), { recursive: true });
    const result = await runProcess(this.executable, [
      "-s",
      this.serial,
      "exec-out",
      "screencap",
      "-p",
    ]);
    await writeFile(output, extractPng(result.stdout));
    return output;
  }
}

File: src/jmgo_controller/adb.py
from __future__ import annotations

import shutil
import subprocess
from pathlib import Path


class AdbError(RuntimeError):
    pass


class Adb:
    def __init__(self, host: str, port: int = 5555, executable: str = "adb"):
        self.host = host
        self.port = port
        resolved = shutil.which(executable)
        if not resolved:
            raise AdbError(
                "adb was not found; install Android Platform Tools "
                "(macOS: brew install --cask android-platform-tools)"
            )
        self.executable = resolved
        self.serial = f"{host}:{port}"

    def _run(
        self, args: list[str], *, capture: bool = True, check: bool = True
    ) -> subprocess.CompletedProcess[str]:
        command = [self.executable, "-s", self.serial, *args]
        try:
            return subprocess.run(
                command,
                check=check,
                text=True,
                capture_output=capture,
            )
        except subprocess.CalledProcessError as error:
            detail = (error.stderr or error.stdout or str(error)).strip()
            raise AdbError(detail) from error

    def connect(self) -> None:
        result = subprocess.run(
            [self.executable, "connect", self.serial],
            check=False,
            text=True,
            capture_output=True,
        )
        output = (result.stdout + result.stderr).lower()
        if result.returncode or "unable" in output or "failed" in output:
            raise AdbError(output.strip())

    def shell(self, *args: str) -> str:
        self.connect()
        return self._run(["shell", *args]).stdout.strip()

    def info(self) -> dict[str, str]:
        values = self.shell(
            "sh",
            "-c",
            "printf '%s\\n' \"$(getprop ro.product.model)\" "
            '"$(getprop ro.build.version.release)" "$(getprop ro.build.version.sdk)" '
            '"$(getprop ro.product.cpu.abilist)"',
        ).splitlines()
        values += [""] * (4 - len(values))
        return dict(zip(("model", "android", "sdk", "abis"), values, strict=True))

    def current_app(self) -> str:
        return self.shell("sh", "-c", "dumpsys activity activities | grep -m 1 mResumedActivity")

    def audio(self) -> str:
        return self.shell(
            "sh",
            "-c",
            "dumpsys audio | grep -E -m 8 'STREAM_MUSIC|Current:|Devices:'",
        )

    def packages(self, query: str | None = None) -> list[str]:
        packages = [
            line.removeprefix("package:")
            for line in self.shell("pm", "list", "packages").splitlines()
        ]
        if query:
            query = query.casefold()
            packages = [package for package in packages if query in package.casefold()]
        return sorted(packages)

    def install(self, files: list[Path], replace: bool = True) -> str:
        if not files:
            raise AdbError("at least one APK is required")
        missing = [str(path) for path in files if not path.is_file()]
        if missing:
            raise AdbError("APK not found: " + ", ".join(missing))
        self.connect()
        command = ["install-multiple" if len(files) > 1 else "install"]
        if replace:
            command.append("-r")
        command.extend(str(path.resolve()) for path in files)
        return self._run(command).stdout.strip()

    def uninstall(self, package: str, keep_data: bool = False) -> str:
        self.connect()
        command = ["uninstall"]
        if keep_data:
            command.append("-k")
        command.append(package)
        return self._run(command).stdout.strip()

    def launch(self, package: str) -> str:
        return self.shell("monkey", "-p", package, "-c", "android.intent.category.LAUNCHER", "1")

    def screenshot(self, destination: Path) -> Path:
        self.connect()
        destination = destination.resolve()
        destination.parent.mkdir(parents=True, exist_ok=True)
        process = subprocess.run(
            [self.executable, "-s", self.serial, "exec-out", "screencap", "-p"],
            check=False,
            capture_output=True,
        )
        if process.returncode:
            raise AdbError(process.stderr.decode(errors="replace").strip())
        destination.write_bytes(process.stdout)
        return destination
